src/EntryPoint.sol

```diff
@@ -168,113 +139,101 @@ contract EntryPoint is EIP712, UUPSUpgradeable, Ownable, ReentrancyGuard {
     // Main
     ////////////////////////////////////////////////////////////////////////
 
+    /// @dev Executes a single encoded user operation.
+    /// `encodedUserOp` is given by `abi.encode(userOp)`, where `userOp` is a struct of type `UserOp`.
+    /// If sufficient gas is provided, returns an error selector that is non-zero
+    /// if there is an error during the payment, verification, and call execution.
+    function execute(bytes calldata encodedUserOp) public payable virtual returns (bytes4 err) {
+        UserOp calldata u;
+        assembly ("memory-safe") {
+            let t := calldataload(encodedUserOp.offset)
+            u := add(t, encodedUserOp.offset)
+            if or(shr(64, t), lt(encodedUserOp.length, 0x20)) { revert(0x00, 0x00) }
+        }
+        uint256 g = u.combinedGas;
+        uint256 gStart = gasleft();
+        uint256 paymentAmount;
+        assembly ("memory-safe") {
+            // Check if there's sufficient gas left for the gas-limited self calls
+            // via the 63/64 rule. This is for gas estimation. If the total amount of gas
+            // for the whole transaction is insufficient, revert.
+            if or(lt(shr(6, mul(gas(), 63)), add(g, _INNER_GAS_OVERHEAD)), shr(64, g)) {
+                mstore(0x00, 0x1c26714c) // `InsufficientGas()`.
+                revert(0x1c, 0x04)
+            }
+
+            let m := mload(0x40) // Grab the free memory pointer.
+            // Copy the encoded user op to the memory to be ready to pass to the self call.
+            calldatacopy(add(m, 0x40), encodedUserOp.offset, encodedUserOp.length)
+            let s := add(m, 0x1c) // Start of the calldata in memory to pass to the self call.
+            let n := add(encodedUserOp.length, 0x44) // Length of the calldata to the self call.
+
+            // To prevent griefing, we need to do two non-reverting gas-limited calls.
+            // Even if the verify and call fails, which the gas will be burned, 
+            // the payment has already been made and can't be reverted.
```

Comment:
I don't understand, why? I was thnking as I write below "you execute, measure the gas executed, multiply it by the exchange rate, and pay in erc20 only the exact amount needed at the end"